brian2/groups/group.py

```diff
@@ -940,6 +940,20 @@ def run_regularly(self, code, dt=None, clock=None, when='start',
         if dt is None and clock is None:
             clock = self._clock
 
+        # Subgroups are normally not included in their parent's
+        # contained_objects list, since there's no need to include them in the
+        # network (they don't perform any computation on their own). However,
+        # if a subgroup declares a `run_regularly` operation, then we want to
+        # include this operation automatically, i.e. with the parent group
+        # (adding just the run_regularly operation to the parent group's
+        # contained objects would no be enough, since the CodeObject needs a
+        # reference to the group providing the context for the operation, i.e.
+        # the subgroup instead of the parent group. See github issue #922
+        source_group = getattr(self, 'source', None)
+        if source_group is not None:
+            if not self in source_group.contained_objects:
+                source_group.contained_objects.append(self)
+
```

Comment:
This feels a bit hacky and prone to reintroduce the same error in future. What about we just always add subgroups to ``contained_objects`` at the point of creating them, but add a new attribute/mixin class ``object_container_only`` or ``ObjectContainerOnly`` that tells the run scheduler to only include the contained objects of this object, but not the object itself?

Reply:
If we add a weakref to ``contained_objects`` that could work? We could also cull weakrefs that no longer point to anything occasionally?

Reply:
The problem with using weakrefs is that it does not work for code like the following (see #922):
```Python
neurons[0:5].run_regularly("v += 1", dt=10 * ms)
```

Reply:
Another idea: when we create a subgroup, we set the ``contained_objects`` list to be the source ``NeuronGroup.contained_object`` list?

Reply:
This is something I tried but it does not quite work: code objects like the `CodeRunner` that is created by `run_regularly` keep a reference to their "owner". However, to avoid circular dependencies, we store these only as weak references. If the `run_regularly` only adds itself to the `contained_object` list of the parent `NeuronGroup`, then the `Subgroup` will be garbage collected and we'll get a "weakly-referenced object no longer exists" during code generation.

Reply:
OK so how about: ``run_regularly`` has a full ref to ``Subgroup``, but ``run_regularly`` is inserted into the parent ``NeuronGroup.contained_objects`` (possibly by having ``Subgroup.contained_objects`` be a computed attribute that picks out ``parent.contained_objects``). Would that work? Feels like that keeps the needed objects alive, but avoids circular references?

Reply:
In principle this should work, I think. But it would mean that we have to make a difference between a subgroup (strong reference) and a standard group (weak reference) -- not sure that's cleaner than the solution in this PR? It would also mean we'd have to change things in `CodeRunner` which is widely used (there's no "run_regularly class").